vyvolavac/Vyvolavac.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/python
+"""
+    Vyvolavac.py: Randomly selects a student with a certain probability.
+    1st argument input_path - input path of student names with (or without) statistics, input file should contain
+    one student per line and the statistics are tab delimited
+    2nd argument (optional): output_path  - output path for new statistics  
+"""
+
+__maintainer__ = "Dagy Tran"
+__version__ = "0.1 beta"
+__status__ = "Development"
+
+
+import sys
+from random import randint
+import matplotlib.pyplot as plt
+
+
+def load_statistics(input_path):
+    stats = {}
+    should_reset = False
+    with open(input_path, 'r') as input_file:
+        for line in input_file:
+            split_line = line.split('\t')
+
+            if len(split_line) == 1:
+                stats[split_line[0].strip()] = (0, 0)
+                should_reset = True
+            elif len(split_line) == 3:
+                stats[split_line[0]] = (int(split_line[1]), int(split_line[2]))
+
+    if should_reset:
+        stats = reset_stats(stats, 2)
+
+    return stats
+
+
+def reset_stats(stats_dict, multiplier=2):
+    number_of_students = len(stats_dict)
+
+    for key, value in stats_dict.items():
+        stats_dict[key] = (0, multiplier * number_of_students)
+    return stats_dict
+
+
+def save_statistics(output_path, stats_dictionary):
+    with open(output_path, 'w') as output_file:
+        lines = list(map(lambda kv: '{}\t{}\t{}'.format(kv[0], kv[1][0], kv[1][1]), stats_dictionary.items()))
+        output_file.write('\n'.join(lines))
+
+
+def redistribute(stats_dictionary, key):
+    new_stats = {}
+    number_of_students = len(stats_dictionary)
+    if key not in stats_dictionary:
+        print('Error: User not in dictionary.')
+        return stats_dictionary
+
+    student_points = stats_dictionary[key][1]
+    if student_points < number_of_students - 1:
+        print('Warning: User {} has not enough points to redistribute. Points {}. Returning unchanged stats'.format(key,
+                                                                                                                    student_points))
+        return stats_dictionary
+    else:
+        for k, (count, points) in stats_dictionary.items():
+            if k != key:
+                new_stats[k] = (count, points + 1)
+            else:
+                new_stats[k] = (count + 1, points - (number_of_students - 1))
+
+    return new_stats
+
+
+def select_student(stats_dict):
+    cumulative_sum = 0
+    for key, (count, points) in stats_dict.items():
+        cumulative_sum += points
+
+    selected_student = ""
+    selected_index = randint(0, cumulative_sum)
+    prefix_sum = 0
+    for key, (count, points) in stats_dict.items():
```

Comment:
Ve starší verzi Pythonu <= 3.5 bys neměl zaručené pořadí elementů v dictionary. Je to implementační detail, na který není dobré spoléhat. Použil bych jinou datovou strukturu - list.

https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-compactdict

Reply:
Tady není potřeba mít zaručené pořadí elementů. Jde o velikost intervalu, do kterého se může ten randint trefit a ty velikosti intervalů jsou tady pevné. Když mám random number a použiju to na random uspořádané pole, výjde z toho stejně random :D.

Reply:
Chápu to tak, že to má fungovat následovně:
Dejme tomu, že máme 3 studenty a mají body 4, 2, 14. Tak 
1. student má interval 1-4
2. student má interval 5-6
3. student má interval 7-20

randint hodí třeba č. 5, tak podle mě by se měl vybrat ten 2. student. Ale pokud tvůj algoritmus by neřešil pořadí a v 1. cyklu by přičetl do prefix_sum 14 bodů, protože náhodou vybral 3. studenta, tak by funkce vybrala toho 3. studenta místo toho 2.

Jak říkáš ty, tak bys v tom případě nemusel dělat tento algoritmus a mohl bys rovnou náhodně někoho vybrat, ale to bychom pak ztratili tu hlavní myšlenku vyvolávače.

Reply:
Možná ve výsledku ten náhodný výběr vyjde nastejno, ale potom ten algoritmus na první pohled není jasný. Ideální by bylo napsat unit test, co se má očekávat, když randint vrátí nějakou hodnotu pro nějaký dict :P , ale byl bych rád, kdyby se k tomuto vyjádřili i ostatní.

Reply:
Chápu, co tím myslíš :). Ono jde právě stejně o nedeterministický algoritmus. Přinejhorším ta část s dict tu nedeterminickou vlastnost podpoří, což není na škodu. Ale jak říkáš, děje se to jen na starších verzích, které používat nebudeme (nebo zatím nevidím důvod, proč bychom měli).